src/peft/tuners/lora/bnb.py

```diff
@@ -56,34 +56,109 @@ def __init__(
             self.update_layer(adapter_name, r, lora_alpha, lora_dropout, init_lora_weights)
             self.active_adapter = adapter_name
 
+        def merge(self):
+            if self.active_adapter not in self.lora_A.keys():
+                return
+            if self.merged:
+                warnings.warn("Already merged. Nothing to do.")
+                return
+            if self.r[self.active_adapter] > 0:
+                warnings.warn(
+                    "Merge lora module to 8-bit linear may get different generations due to rounding errors."
+                )
+                lora_data = self.get_delta_weight(self.active_adapter)
+
+                if self.state.SCB is None:
+                    self.state.SCB = self.weight.SCB
+                # Dequantize the result of identify matrix and int8 weight because bitsandbytes only have this method.
+                im = torch.eye(self.weight.data.shape[-1]).contiguous().half().to(self.weight.device)
+                im, imt, SCim, SCimt, coo_tensorim = bnb.functional.double_quant(im)
+                im, Sim = bnb.functional.transform(im, "col32")
+
+                if self.state.CxB is None:
+                    self.state.CxB, self.state.SB = bnb.functional.transform(
+                        self.weight.data, to_order=self.state.formatB
+                    )
+                out32, Sout32 = bnb.functional.igemmlt(im, self.state.CxB, Sim, self.state.SB)
+                output = bnb.functional.mm_dequant(out32, Sout32, SCim, self.state.SCB, bias=None).t()
+                w_data = output.to(lora_data.dtype).to(lora_data.device) + lora_data
+                self.weight = bnb.nn.Int8Params(
+                    w_data.to("cpu"), requires_grad=False, has_fp16_weights=self.weight.has_fp16_weights
+                ).to(self.weight.device)
+                self.state.reset_grads()
+                self.merged = True
+
+        def unmerge(self):
```

Comment:
Could you please extend the test, or add a separate one, to also test unmerging?

Reply:
I think `test_8bit_merge_and_disable_lora` in `test_common_gpu.py` is for testing unmerge since `model.disable_adapter()` will call `unmerge()` in the [forward](https://github.com/huggingface/peft/blob/main/src/peft/tuners/lora/bnb.py#L170).
Do I misunderstand it?

Reply:
Yes, you are right, sorry I missed that.

Reply:
np.

I have fixed the annotation and removed `"default"` in `merge_and_unload()` in the tests, would you please help me review these? Thx!